refactor(cwg): switch to external radius package

wolfseb opened the pull request:
## Summary

Closes #14184 and #13671.

This PR replaces the local fork of the radius package with `github.com/layeh/radius`.
Additionally, packages that were only used for the xwf that were linked to radius are removed (Ruckus and Expresswifi)

Since the packet `radius/rfc2548` does not exist anymore we defined the MicrosoftVendor specific variables (MSMPPE key types) explicitly.

The most significant changes in functionality is that the type `radius.Attributes` changes from a `map{Type: []Attribute}` to a list of pointers: `[]*Attributes`, where `Attributes` is a struct containing the types `Type` and `Attribute`.

## Test Plan

Run all CWF unit and integration tests.

Local CWF runner: https://github.com/wolfseb/magma/actions/runs/3334873921

## Additional Information

- [ ] This change is backwards-breaking

Created in pairing with @MoritzThomasHuebner

MoritzThomasHuebner committed:
refactor(feg): Removed ruckus

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
refactor(feg): remove expresswifi

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(feg): Switch to layeh.com/radius

Removes expresswifi and switches from local Radius library to layeh.com/radius

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(feg): remove local radius fork

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

crasu commented on feg/radius/src/modules/coadynamic/coa_dynamic_test.go:
```diff
@@ -36,22 +39,16 @@ func TestCoaDynamic(t *testing.T) {
-		return
-	}
-	fmt.Println("Server listenning")
+	time.Sleep(10 * time.Millisecond)
```
Would be nice to do this without a sleep

wolfseb replied:
Done.

crasu commented on feg/radius/src/modules/coafixedip/coa_fixed_ip_test.go:
```diff
@@ -46,17 +49,13 @@ func TestCoaFixed(t *testing.T) {
-	if !listenSuccess {
-		require.Fail(t, "radiusServer start error")
-	}
+	time.Sleep(10 * time.Millisecond)
```
please no new sleeps

wolfseb replied:
Done.

crasu commented on feg/radius/src/modules/coanas/coa_nas_test.go:
```diff
@@ -39,18 +41,13 @@ func TestCoaNas(t *testing.T) {
-		require.Fail(t, "radiusServer start error")
-		return
-	}
+	time.Sleep(10 * time.Millisecond)
```
replace this one  as well

wolfseb replied:
Done.

crasu commented on feg/radius/src/modules/coanas/coa_nas_test.go:
```diff
@@ -98,18 +95,13 @@ func TestCoaNasNoResponse(t *testing.T) {
-		require.Fail(t, "radiusServer start error")
-		return
-	}
+	time.Sleep(10 * time.Millisecond)
```
and another one

wolfseb replied:
Done.

crasu commented on feg/radius/src/modules/coanas/coa_nas_test.go:
```diff
@@ -153,18 +145,13 @@ func TestCoaNasFieldInvalid(t *testing.T) {
-		require.Fail(t, "radiusServer start error")
-		return
-	}
+	time.Sleep(10 * time.Millisecond)
```
see above

wolfseb replied:
Done.

Commit:
refactor(feg): replace sleeps in unit tests

Replace the sleeps to wait for the radius server to be active with
exchanging packages until the server responds.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner commented on feg/radius/src/filters/lballocate/lb_allocate.go:
```diff
@@ -15,13 +15,15 @@ package lballocate
-	"math/rand"
-	"time"
+
+	"layeh.com/radius"
```
There should not be an empty line after the radius import.

MoritzThomasHuebner commented on feg/radius/src/filters/lballocate/lb_allocate_test.go:
```diff
@@ -15,15 +15,17 @@ package lballocate
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+
```
There should also be no empty lines between the external library imports.

MoritzThomasHuebner commented on feg/radius/src/filters/lbcanary/lb_canary.go:
```diff
@@ -15,13 +15,15 @@ package lbcanary
-	"math/rand"
-	"time"
+
+	"layeh.com/radius"
```
same here

MoritzThomasHuebner commented on feg/radius/src/filters/lbcanary/lb_canary_test.go:
```diff
@@ -15,13 +15,15 @@ package lbcanary
-	"testing"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
same here

MoritzThomasHuebner commented on feg/radius/src/loader/static_loader.go:
```diff
@@ -33,9 +33,7 @@ import (
-	modxwfv3 "fbc/cwf/radius/modules/xwfv3"
-	"fbc/lib/go/radius"
-	"fmt"
+	"layeh.com/radius"
```
Same here, layeh.com/radius should be with zap

MoritzThomasHuebner commented on feg/radius/src/modules/alwaysaccept/always_accept.go:
```diff
@@ -15,8 +15,10 @@ package alwaysaccept
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
same here

MoritzThomasHuebner commented on feg/radius/src/modules/alwaysaccept/always_accept_test.go:
```diff
@@ -14,10 +14,12 @@ limitations under the License.
+	"fbc/cwf/radius/modules"
+
+	"layeh.com/radius"
+
```
same here

MoritzThomasHuebner commented on feg/radius/src/modules/analytics/analytics.go:
```diff
@@ -24,10 +24,11 @@ import (
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+	"layeh.com/radius/rfc2869"
```
same here

MoritzThomasHuebner commented on feg/radius/src/modules/analytics/xwf_vendors.go:
```diff
@@ -20,7 +20,6 @@ type Vendor int
 // Enum list of vendors supported by Express-Wifi program in its wider meaning (Wifi, Carrier Wifi, Standalone, ...).
 const (
```
Do we need this file at all at this point?

wolfseb replied:
No, we don't. Good catch!

MoritzThomasHuebner commented on feg/radius/src/modules/coadynamic/coa_dynamic.go:
```diff
@@ -3,12 +3,14 @@ package coadynamic
+	"fbc/cwf/radius/modules/coadynamic/radiustracker"
+
+	"layeh.com/radius"
+
```
external library imports should not have empty lines between them.

MoritzThomasHuebner commented on feg/radius/src/modules/coadynamic/coa_dynamic_test.go:
```diff
@@ -2,14 +2,16 @@ package coadynamic
 
+	"fbc/cwf/radius/modules"
+
+	"layeh.com/radius"
```
same here

MoritzThomasHuebner commented on feg/radius/src/modules/coadynamic/coa_dynamic_test.go:
```diff
@@ -70,6 +68,16 @@ func TestCoaDynamic(t *testing.T) {
 }
 
+func waitForRadiusServerToBeReady(secret []byte, port int) {
+	MaxRetries := 10
```
Should we return an error in this function? How would we be able to tell if there was an issue with `radius.Exchange`. Previously, we handled possible errors in `TestCoaDynamic`.

wolfseb replied:
Good point, I made it return an error if it exceeds the number of retries and assert that the returned error is nil.

wolfseb replied:
The point of `radius.exchange` is that it returns an error as long as the radius server is not yet running, and the function just retries the exchanges a couple of times until the server is up and responsive.

MoritzThomasHuebner commented on feg/radius/src/modules/coafixedip/coa_fixed_ip.go:
```diff
@@ -22,7 +22,8 @@ import (
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/coafixedip/coa_fixed_ip_test.go:
```diff
@@ -15,11 +15,13 @@ package coafixedip
 
+	"fbc/cwf/radius/modules"
+
+	"layeh.com/radius"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/coafixedip/coa_fixed_ip_test.go:
```diff
@@ -80,6 +79,16 @@ func TestCoaFixed(t *testing.T) {
 }
 
+func waitForRadiusServerToBeReady(secret []byte, port int) {
+	MaxRetries := 10
```
This function should not exist three times. Maybe you can move it to where `"fbc/cwf/radius/modules"` is, since this is already imported in all the coa test files.

wolfseb replied:
It is now moved to a central place at `fbc/cwf/radius/modules/modules_test_utils.go`, where several tests in the folder will call it from.

MoritzThomasHuebner commented on feg/radius/src/modules/coanas/coa_nas_test.go:
```diff
@@ -188,6 +174,16 @@ func TestCoaNasFieldInvalid(t *testing.T) {
 }
 
+func waitForRadiusServerToBeReady(secret []byte, port int) {
+	MaxRetries := 10
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/authstate/manager.memory.go:
```diff
@@ -15,15 +15,16 @@ package authstate
 
-	"fbc/lib/go/radius"
-	"fbc/lib/go/radius/rfc2865"
+	"layeh.com/radius"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/authstate/manager.memory_test.go:
```diff
@@ -14,14 +14,15 @@ limitations under the License.
+	"fbc/cwf/radius/modules/eap/packet"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/eap.go:
```diff
@@ -15,15 +15,17 @@ package eap
-	"fmt"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2869"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/methods/akamagma/aka_magma.go:
```diff
@@ -24,8 +24,9 @@ import (
-	"fbc/lib/go/radius/rfc2865"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/methods/akatataipx/akatataipx.go:
```diff
@@ -11,8 +11,9 @@ import (
-	"fbc/lib/go/radius/rfc2865"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/eap/methods/common/keying.go:
```diff
@@ -16,27 +16,32 @@ package common
 var MicrosoftVendor = []byte{0x00, 0x00, 0x01, 0x37}
 
+const (
+	MSMPPESendKey_Type = radius.Type(16)
```
We should add a comment line about where these constants come from

MoritzThomasHuebner commented on feg/radius/src/modules/lbserve/lb_serve.go:
```diff
@@ -18,7 +18,8 @@ import (
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/lbserve/lb_serve_test.go:
```diff
@@ -15,13 +15,16 @@ package lbserve
+	"fbc/cwf/radius/session"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/lbserve/lb_serve_test.go:
```diff
@@ -142,13 +145,12 @@ func spawnRadiusServer() (server *radius.PacketServer, port int) {
 		_ = server.ListenAndServe()
 	}()
-	<-server.Ready // Wait for server to get ready
+	time.Sleep(10 * time.Millisecond)
```
Should we also try to eliminate this sleep?

MoritzThomasHuebner commented on feg/radius/src/modules/lbserve/lb_serve_test.go:
```diff
@@ -206,21 +208,20 @@ func spawnFailingRadiusServer() (server *radius.PacketServer, port int) {
 		_ = server.ListenAndServe()
 	}()
-	<-server.Ready // Wait for server to get ready
+	time.Sleep(10 * time.Millisecond)
```
see above

MoritzThomasHuebner commented on feg/radius/src/modules/lbserve/lb_serve_test.go:
```diff
@@ -206,21 +208,20 @@ func spawnFailingRadiusServer() (server *radius.PacketServer, port int) {
-	packet.Attributes[rfc2865.CallingStationID_Type] = []radius.Attribute{radius.Attribute("called")}
-	packet.Attributes[rfc2865.CalledStationID_Type] = []radius.Attribute{radius.Attribute("calling")}
+	packet.Add(rfc2865.CallingStationID_Type, radius.Attribute("calling"))
+	packet.Add(rfc2865.CalledStationID_Type, radius.Attribute("called"))
```
We are turning `calling` and `called` around here. Are we sure this is correct? At the least this should be mentioned in the PR summary.

wolfseb replied:
I changed it back to how it was originally. Everywhere else these attributes are switched, but in the end it doesn't matter at all what's written here, as long as the attributes are not `nil`, as that's the only thing that's asserted in the tests..

MoritzThomasHuebner commented on feg/radius/src/modules/magmaacct/magma_acct.go:
```diff
@@ -23,10 +23,11 @@ import (
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+	"layeh.com/radius/rfc2869"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/module.go:
```diff
@@ -22,7 +22,8 @@ import (
 	"fbc/cwf/radius/session"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/modulestest/module_mock.go:
```diff
@@ -15,7 +15,8 @@ package modulestest
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/ofpanalytics/ofpanalytics.go:
```diff
@@ -24,8 +24,9 @@ import (
-	"fbc/lib/go/radius/rfc2865"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/proxy/proxy.go:
```diff
@@ -18,7 +18,8 @@ import (
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/proxy/proxy_test.go:
```diff
@@ -15,13 +15,16 @@ package proxy
+	"fbc/cwf/radius/session"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/proxy/proxy_test.go:
```diff
@@ -51,17 +54,13 @@ func TestProxy(t *testing.T) {
-	if !listenSuccess {
-		return
-	}
+	time.Sleep(10 * time.Millisecond)
```
See above, can we eliminate the sleep?

MoritzThomasHuebner commented on feg/radius/src/modules/testloopback/test_loopback.go:
```diff
@@ -15,7 +15,8 @@ package testloopback
 	"fbc/cwf/radius/modules"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/modules/testsessionstorage/testsessionstorage.go:
```diff
@@ -16,7 +16,8 @@ package testsessionstorage
 	"fbc/cwf/radius/session"
-	"fbc/lib/go/radius"
+
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/monitoring/listener.go:
```diff
@@ -2,10 +2,11 @@ package monitoring
 	"fmt"
 	"sync"
 
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/server/grpc_listener.go:
```diff
@@ -16,18 +16,19 @@ package server
-	"fbc/lib/go/radius/rfc2866"
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
```
See above

MoritzThomasHuebner commented on feg/radius/src/server/server.go:
```diff
@@ -28,7 +28,7 @@ import (
 	"go.opencensus.io/tag"
 
-	"fbc/lib/go/radius"
+	"layeh.com/radius"
```
See above

MoritzThomasHuebner commented on feg/radius/src/server/server_test.go:
```diff
@@ -31,10 +31,11 @@ import (
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+	"layeh.com/radius/rfc2869"
```
See above

MoritzThomasHuebner commented on feg/radius/src/server/udp_listener.go:
```diff
@@ -22,10 +22,10 @@ import (
-
 	"fbc/cwf/radius/session"
 
+	"layeh.com/radius"
```
See above

Commit:
chore(feg): cleanup go imports and leftovers

* Cleanup go import order and groupings
* Remove leftover xwf vendors file
* Introduce error handling for WaitForRadiusServerToBeReady function and
move it to a central place
* replace more sleeps with the new wait function

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
fix(feg): remove xwf specific vendors

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

crasu reviewed (commented):
LGTM

sebathomas commented on feg/radius/src/modules/lbserve/lb_serve_test.go:
```diff
@@ -15,17 +15,17 @@ package lbserve
 
```
This unit test (and a few others) fail when I try to run them on my machine. It might be a problem on my machine though. Do you know if the unit tests run in the CI?

![Screenshot_20221102_163439](https://user-images.githubusercontent.com/14236667/199534063-4f16cdea-28c5-4ad1-9f4a-114fe399ffd2.png)

wolfseb replied:
They should work again now (see comment below)

sebathomas replied:
They're now green for me too. But it seems the CI did not notice the failures before.

I'm looking at that in #14347.

sebathomas replied:
I created a follow-up task for that:
#14350

sebathomas commented on feg/radius/src/modules/module_test_utils.go:
```diff
@@ -0,0 +1,33 @@
+			return nil
+		}
+	}
+	return errors.New(fmt.Sprintf("radius server failed to be ready after %d retries: %v", MaxRetries, err))
```
Some tests fail with this error message for me. Do we need to add some kind of sleep statement to the retries?

wolfseb replied:
For me it always worked before, but I get the same error now. The problem was that for the server address, the port number must come after the last colon, i.e. if it's just a port it must be prepended with a colon. I fixed it and the unit tests should now work again.

sebathomas commented on feg/radius/src/server/udp_listener.go:
```diff
@@ -86,15 +83,10 @@ func (l *UDPListener) ListenAndServe() error {
 
-	// Wait to see if initialization was successful
-	select {
-	case _ = <-l.Server.Ready:
```
Did we need to remove the `ready` flag because it's not in the external radius package?

wolfseb replied:
Yes, exactly. The external package has no way as far as I can tell to check for the server to be ready, other than sending something and seeing if it came back with an error (that is also how I constructed the wait function for the unit tests)

wolfseb replied:
We could replace it maybe with a wait function similar to the one in the tests, instead of just hoping that the server comes up quickly enough

Commit:
fix(feg): fix address in radius unit tests

The address passed to radius needs to have a colon before the port.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
